pkg/assembler/backends/testing/osv.go

```diff
@@ -17,77 +17,181 @@ package testing
+func (n *osvNode) getID() uint32   { return n.id }
+
+// Ingest OSV
+func (c *demoClient) IngestOsv(ctx context.Context, input *model.OSVInputSpec) (*model.Osv, error) {
+	osvStruct, hasOsv := c.osvs[osv]
+	if !hasOsv {
+		osvStruct = &osvNode{
+			id:     c.getNextID(),
+			osvIDs: osvIDMap{},
 		}
+		c.index[osvStruct.id] = osvStruct
+		c.osvs[osv] = osvStruct
 	}
+	osvIDs := osvStruct.osvIDs
+	osvID := strings.ToLower(input.OsvID)
 
-	o.OsvID = append(o.OsvID, &model.OSVId{ID: id})
-	return o
+	osvIDStruct, hasOsvID := osvIDs[osvID]
+	if !hasOsvID {
+		osvIDStruct = &osvIDNode{
+			id:     c.getNextID(),
+			parent: osvStruct.id,
+			osvID:  osvID,
+		}
+		c.index[osvIDStruct.id] = osvIDStruct
+		osvIDs[osvID] = osvIDStruct
+	}
+
+	// build return GraphQL type
+	return c.buildOsvResponse(osvIDStruct.id, nil)
 }
 
 // Query OSV
+func (c *demoClient) Osv(ctx context.Context, filter *model.OSVSpec) ([]*model.Osv, error) {
+	if filter != nil && filter.ID != nil {
+		id, err := strconv.Atoi(*filter.ID)
+		if err != nil {
+			return nil, err
+		}
+		osv, err := c.buildOsvResponse(uint32(id), filter)
+		if err != nil {
+			return nil, err
+		}
+		return []*model.Osv{osv}, nil
+	}
+	out := []*model.Osv{}
+	for _, osvNode := range c.osvs {
+		osvIDList := []*model.OSVId{}
+		if filter != nil && filter.OsvID != nil {
+			osvIDNode, hasOsvIDNode := osvNode.osvIDs[strings.ToLower(*filter.OsvID)]
+			if hasOsvIDNode {
+				osvIDList = append(osvIDList, &model.OSVId{
+					ID:    fmt.Sprintf("%d", osvIDNode.id),
+					OsvID: osvIDNode.osvID,
+				})
+			}
+		} else {
+			for _, osvIDNode := range osvNode.osvIDs {
+				osvIDList = append(osvIDList, &model.OSVId{
+					ID:    fmt.Sprintf("%d", osvIDNode.id),
+					OsvID: osvIDNode.osvID,
+				})
+			}
+		}
+		if len(osvIDList) > 0 {
+			out = append(out, &model.Osv{
+				ID:     fmt.Sprintf("%d", osvNode.id),
+				OsvIds: osvIDList,
+			})
+		}
+	}
+	return out, nil
+}
 
-func (c *demoClient) Osv(ctx context.Context, osvSpec *model.OSVSpec) ([]*model.Osv, error) {
-	var osv []*model.Osv
-	for _, o := range c.osv {
-		newOSV, err := filterOSVID(o, osvSpec)
+// Builds a model.osv to send as GraphQL response, starting from id.
+// The optional filter allows restricting output (on selection operations).
+func (c *demoClient) buildOsvResponse(id uint32, filter *model.OSVSpec) (*model.Osv, error) {
+	if filter != nil && filter.ID != nil {
+		filteredID, err := strconv.Atoi(*filter.ID)
 		if err != nil {
 			return nil, err
 		}
-		if newOSV != nil {
-			osv = append(osv, newOSV)
+		if uint32(filteredID) != id {
+			return nil, nil
+		}
+	}
+
+	node, ok := c.index[id]
+	if !ok {
+		return nil, gqlerror.Errorf("ID does not match existing node")
+	}
+
+	osvIDList := []*model.OSVId{}
+	if osvIDNode, ok := node.(*osvIDNode); ok {
+		if filter != nil && noMatchLowerCase(filter.OsvID, osvIDNode.osvID) {
+			return nil, nil
 		}
+		osvIDList = append(osvIDList, &model.OSVId{
+			ID:    fmt.Sprintf("%d", osvIDNode.id),
```

Comment:
nit: may be useful to make this a helper function in case we decide to change it and also in case someone puts a slightly different format string that may lead to something different